src/parse.c

```diff
@@ -1648,6 +1648,184 @@ handle_tunnel_key_mapping(yaml_document_t* doc, yaml_node_t* node, const void* _
+    if (strlen(endpoint) < 3) {
+        return yaml_error(node, error, "invalid endpoint address or hostname '%s'", scalar(node));
+    }
+    if (endpoint[0] == '[') {
+        /* this is an ipv6 endpoint in [ad:rr:ee::ss]:port form */
+        char *endbrace = strrchr(endpoint, ']');
+        if (!endbrace)
+            return yaml_error(node, error, "invalid address in endpoint '%s'", scalar(node));
+        address = endpoint + 1;
+        *endbrace = '\0';
+        port = strrchr(endbrace + 1, ':');
+    } else {
+        address = endpoint;
+        port = strrchr(endpoint, ':');
+    }
+    /* split off :port */
+    if (!port)
+        return yaml_error(node, error, "endpoint '%s' is missing :port", scalar(node));
+    *port = '\0';
+    port++; /* skip former ':' into first char of port */
+    port_num = g_ascii_strtoull(port, NULL, 10);
+    if (port_num > 65535)
+        return yaml_error(node, error, "invalid port in endpoint '%s'", scalar(node));
+    if (is_ip4_address(address) || is_ip6_address(address) || is_hostname(address)) {
+        return handle_wireguard_peer_str(doc, node, &peer->endpoint, error);
+    }
+    return yaml_error(node, error, "invalid endpoint address or hostname '%s'", scalar(node));
+}
+
+#define wireguard_peer_offset(field) GUINT_TO_POINTER(offsetof(wireguard_peer, field))
+const mapping_entry_handler wireguard_peer_handlers[] = {
+    {"public-key", YAML_SCALAR_NODE, handle_wireguard_peer_str, NULL, wireguard_peer_offset(public_key)},
+    {"preshared-key", YAML_SCALAR_NODE, handle_wireguard_peer_str, NULL, wireguard_peer_offset(preshared_key)},
+    {"preshared-key-file", YAML_SCALAR_NODE, handle_wireguard_peer_str, NULL, wireguard_peer_offset(preshared_key_file)},
+    {"keepalive", YAML_SCALAR_NODE, handle_wireguard_peer_guint, NULL, wireguard_peer_offset(keepalive)},
+    {"endpoint", YAML_SCALAR_NODE, handle_wireguard_endpoint},
+    {"allowed-ips", YAML_SEQUENCE_NODE, handle_wireguard_allowed_ips},
+
+    {NULL}
+};
+
+static gboolean
+process_wireguard_peer(yaml_document_t* doc, yaml_node_t* node, wireguard_peer *peer, GError** error)
+{
+    for (yaml_node_pair_t *entry = node->data.mapping.pairs.start; entry < node->data.mapping.pairs.top; entry++) {
+        yaml_node_t* key, *value;
+        const mapping_entry_handler* h;
+
+        g_assert(*error == NULL);
+
+        key = yaml_document_get_node(doc, entry->key);
+        value = yaml_document_get_node(doc, entry->value);
+        assert_type(key, YAML_SCALAR_NODE);
+        h = get_handler(wireguard_peer_handlers, scalar(key));
+        if (!h)
+            return yaml_error(key, error, "unknown key '%s'", scalar(key));
+        assert_type(value, h->type);
+
+        if (!h->handler(doc, value, (void *)peer + GPOINTER_TO_UINT(h->data), error))
+            return FALSE;
+    }
+    return TRUE;
+}
+
+static gboolean
+handle_wireguard_peers(yaml_document_t* doc, yaml_node_t* node, const void* _, GError** error)
```

Comment:
I'd prefer that the Wireguard peers would be handled the same way as `NetplanWifiAccessPoint`/`handle_wifi_acccess_points`, using a global `cur_wireguard_peer` variable.

This way we could avoid storing the whole `NetplanWireGuardPeer`/`wireguard_peer` struct data inside the GArray and store just the pointers to the corresponding data instead.